sdc/datatypes/hpat_pandas_rolling_types.py

```diff
@@ -0,0 +1,241 @@
+class RollingType(types.Type):
+    """Type definition for pandas.rolling functions handling."""
+    def __init__(self, ty, data, win_type=None, on=None, closed=None):
+        self.data = data
+        self.win_type = win_type or types.none
+        self.on = on or types.none
+        self.closed = closed or types.none
+
+        name_tmpl = '{}({}, win_type={}, on={}, closed={})'
+        name = name_tmpl.format(ty, data, self.win_type, self.on, self.closed)
+        super(RollingType, self).__init__(name)
+
+
+class RollingTypeModel(StructModel):
+    """Model for RollingType type."""
+    def __init__(self, dmm, fe_type):
+        members = [
+            ('data', fe_type.data),
+            # window is able to be offset
+            ('window', types.intp),
+            ('min_periods', types.intp),
+            ('center', types.boolean),
+            ('win_type', fe_type.win_type),
+            ('on', fe_type.on),
+            # axis is able to be unicode type
+            ('axis', types.intp),
+            ('closed', fe_type.closed),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+make_attribute_wrapper(RollingType, 'data', '_data')
+make_attribute_wrapper(RollingType, 'window', '_window')
+make_attribute_wrapper(RollingType, 'min_periods', '_min_periods')
+make_attribute_wrapper(RollingType, 'center', '_center')
+make_attribute_wrapper(RollingType, 'win_type', '_win_type')
+make_attribute_wrapper(RollingType, 'on', '_on')
+make_attribute_wrapper(RollingType, 'axis', '_axis')
+make_attribute_wrapper(RollingType, 'closed', '_closed')
+
+
+def gen_hpat_pandas_rolling_init(ty):
+    """Generate rolling initializer based on data type"""
+    def _hpat_pandas_rolling_init(typingctx, self, window, min_periods=None,
+                                  center=False, win_type=None,
+                                  on=None, axis=0, closed=None):
+        """Internal Numba required function to register RollingType."""
+        ret_typ = ty(self, win_type, on, closed)
+        sig = signature(ret_typ, self, window, min_periods,
+                        center, win_type, on, axis, closed)
+
+        def _codegen(context, builder, sig, args):
+            """Create DataFrameRollingTypeModel structure."""
+            data, window, min_periods, center, win_type, on, axis, closed = args
+            rolling = cgutils.create_struct_proxy(sig.return_type)(context, builder)
+            rolling.data = data
+            rolling.window = window
+            rolling.min_periods = min_periods
+            rolling.center = center
+            rolling.win_type = win_type
+            rolling.on = on
+            rolling.axis = axis
+            rolling.closed = closed
+
+            if context.enable_nrt:
+                context.nrt.incref(builder, self, rolling.data)
+
+            return rolling._getvalue()
+
+        return sig, _codegen
+
+    return _hpat_pandas_rolling_init
+
+
+def gen_sdc_pandas_rolling(initializer, ty):
```

Comment:
It's not very clear from the name of the function what it does. I would suggest:
```suggestion
def gen_sdc_pandas_rolling_overload_body(ctor, type):
```

Also the docstring seems to be very obscure. Maybe it would be better to state that it generates code of the overloaded method using associated DataType and constructor.